### PR TITLE
[Merge with CI PR]  Secrets from vault via DynaConf

### PR DESCRIPTION
**Pull request by jyejare**
**The PR integrates Hashicorp Vault as a secret store via DynaConf.**

The PR is dependent on following 2 DynaConf Upstream PRs which are now merged:

- [x] https://github.com/rochacbruno/dynaconf/pull/725 
- [x] https://github.com/rochacbruno/dynaconf/pull/729 

And CI MR (Need to merge along with this PR):
- [x] CI changes (satelliteqe-jenkins !570)

Tests Results:
----------------

Check Critical Tests : https://{stage_url}/job/jyejare-6.11-rhel7-critical/2
Check High Tests: https://{stage_url}/job/jyejare-6.11-rhel7-high/1

### CODE REVIEW

**Comment by jyejare** [comment]
Both the DynaConf PRs are now merged :)

**Comment by jyejare** [comment]
@SatelliteQE/robottelo-tier-1-reviewers and @SatelliteQE/robottelo-tier-2-reviewers The PR is ready for review but hold merging for CI PR.

File: robottelo/config/validators.py
Comment (lhellebr) [inline robottelo/config/validators.py:10]: How would this happen? Wouldn't it only fail for that certain section?

**Review by lhellebr — APPROVED** [review]
This needs to be tested but LGTM

File: robottelo/config/validators.py
Comment (JacobCallahan) [reply robottelo/config/validators.py:10]: validation currently happens on the entire settings object and once a validation error is encountered, it stops

File: robottelo/config/validators.py
Comment (jyejare) [inline robottelo/config/validators.py:68]: @JacobCallahan @mshriver I think we should remove these now ?

Comment (mshriver) [reply robottelo/config/validators.py:68]: There are still references to these secrets within the test framework, can't remove them until we actually remove those tests, which won't be happening anytime soon.

File: .env
Comment (mshriver) [inline .env:23]: I think this should be far more generic for this field.

File: requirements.txt
Comment (mshriver) [inline requirements.txt:8]: Have you requested a release with those updates? I'm hesitant to go with unreleased master branch as our dependency.

File: robottelo/config/validators.py
Comment (mshriver) [inline robottelo/config/validators.py:150]: Thanks for cleaning this up!

File: robottelo/config/validators.py
Comment (mshriver) [inline robottelo/config/validators.py:71]: This is hardcoding all of the vault keys into source here, this is a huge downside to the approach of using validator defaults.

File: .env
Comment (jyejare) [reply .env:23]: Didnt get the comment :) Can you elaborate on this ?

File: robottelo/config/validators.py
Comment (jyejare) [reply robottelo/config/validators.py:71]: Agree with hardcoding but I dont see how that could be a downside? 
The hardcoding in validators is the same as the hardcoded names of setting in validators!

Everyone in the team has access to secrets so shouldn't be an issue to access, and update the secrets.

**Comment by jyejare** [comment]
@lhellebr The test results are available in Stage Jenkins:

Check Critical Tests : https://{stage_url}/job/jyejare-6.11-rhel7-critical/2
Check High Tests: https://{stage_url}/job/jyejare-6.11-rhel7-high/1

File: .env
Comment (mshriver) [reply .env:23]: This file in our upstream open-source repo should not contain the vault path that we use internally.

File: robottelo/config/validators.py
Comment (mshriver) [reply robottelo/config/validators.py:71]: There is a big difference between the two, and its not about access/updating.

The names of the settings map directly to the yaml files and data structure within them. `vmware.image_password` as an argument for the validator is dictating both the structure of the configuration yaml and the acceptable values for a given field.

This structure that you've proposed now hardcodes the paths within vault, in our logic code, instead of within the config files themselves. Users of robottelo shouldn't have to modify this module to not use `satqe_shared_password` as their vault key, and they shouldn't be forced to use a key like `satqe_shared_password`.

Assuming I want to change the credential for this `vmware.image_password` field. It is now hardcoded in source as `satqe_shared_password`, and I can't update vault without impacting every other field mapped to this vault key, and/or updating source code.

This is violating the 12 factor app principles, which is a big reason for why we adopted dynaconf. Config should be cleanly and clearly separated from code, with the only 'overlap' being these validator paths and conf file location, because the config can't be completely arbitrary. 

@JacobCallahan and I discussed the option of trying to use dynaconf dynamic variable formatting for this, using `ssh_password: '@format {this.secrets.ssh_password}'` within the yaml file, and letting the dynaconf instance for `settings` object resolve from its own secrets loading, instead of creating two dynaconf instances. 

This isn't my decision alone to make, and I've thoroughly voiced my issues with this approach and have suggested the approach of using a yamlLoader to proxy these lookups like IQE has done.  If you and @JacobCallahan have heard these warnings and would like to proceed with this implementation, I won't stop you. But I think its setting us up for problems as-is in this PR.

File: .env
Comment (mshriver) [inline .env:2]: There are no docs included guiding users on what to do for local configuration of vault.

This file should be a template, or this block dumped by vault_login script if not present.

The user can't modify this file without source control seeing it. The contents here should be in an example file, so that the user can have a `.env` file locally.

File: robottelo/config/__init__.py
Comment (mshriver) [inline robottelo/config/__init__.py:18]: I think we should continue to explicitly use LazySettings

File: requirements.txt
Comment (mshriver) [reply requirements.txt:8]: I'd say your fork is a better option than master, neither are as good as a release.

These requirements alone don't seem to be giving me enough to run `vault_login` script, as the executable for vault isn't included. 

Are you expecting users to have added the hashicorp repo and installed via DNF? 
https://learn.hashicorp.com/tutorials/vault/getting-started-install

File: .env
Comment (jyejare) [reply .env:2]: Cool! This can be moved to example file, there is no harm though !

File: requirements.txt
Comment (jyejare) [reply requirements.txt:8]: Yes! I am expecting people to install vault themselves, depending on their OS. The installation differs for different OSes.

File: robottelo/config/__init__.py
Comment (jyejare) [reply robottelo/config/__init__.py:18]: Its mentioned here: https://github.com/rochacbruno/dynaconf/blob/master/dynaconf/__init__.py#L19

File: robottelo/config/validators.py
Comment (jyejare) [reply robottelo/config/validators.py:71]: @mshriver RE:

> This structure that you've proposed now hardcodes the paths within vault, in our logic code, instead of within the config files themselves. Users of robottelo shouldn't have to modify this module to not use satqe_shared_password as their vault key, and they shouldn't be forced to use a key like satqe_shared_password.

Users can still use config files to not use secrets from vault, by just updating the setting in conf. infact,  version controlling the vault secret keys with validators will allow us to monitor the secret names changes in vault PnT store and hence we will have better control over modifying secrets from Sat rover group in the vault.

> Assuming I want to change the credential for this vmware.image_password field. It is now hardcoded in source as satqe_shared_password, and I can't update vault without impacting every other field mapped to this vault key, and/or updating source code.

Again, this comes under the version controlling the change in secrets and a better control as mentioned in my previous comment. Updating the vault for specific secret credentials and its impact on other field mapped to the key isnt an issue with the hardcoded approach itself. Its the issue of using same credentials for multiple secrets and thats issue even now with OCP secrets. To resolve it, we can have 1:1 mapping with secret and its credential even though the credentials are the same for multiple secrets.

> This is violating the 12 factor app principles, which is a big reason for why we adopted dynaconf. Config should be cleanly and clearly separated from code, with the only 'overlap' being these validator paths and conf file location, because the config can't be completely arbitrary.

TBH, I am not a fan of principles unless the refactor / approach makes things easier to access, monitor and control the credentials from Vault. 

> @JacobCallahan and I discussed the option of trying to use dynaconf dynamic variable formatting for this, using ssh_password: '@format {this.secrets.ssh_password}' within the yaml file, 

What I dont like the YAML approach is the user has to remember the format and the entire path e.g `this.secrets.ssh_password` in case the user updated the conf locally with some credential string and now wants to revert back so that the user can access secret from the vault. This way we are forcing users to remember the path or every time go and look into CI to pick copy the path to access secrets instead of automation(now Validators) doing it for us. 

> and letting the dynaconf instance for settings object resolve from its own secrets loading, instead of creating two dynaconf instances.

This could be the only reason I would go with the YAML approach but anyways they are two different objects of dynaConf settings and shouldn't be interfering with each other.

Comment (jyejare) [reply robottelo/config/validators.py]: The changes were made as requested after @JacobCallahan convinced me on some security aspects of it!

File: requirements.txt
Comment (mshriver) [reply requirements.txt:8]: 3.1.8 is out (and there's a pyup bot PR available with it) so this change can be dropped.

Comment (mshriver) [reply requirements.txt:8]: @jyejare thanks for pushing for the dynaconf release!

File: robottelo/config/__init__.py
Comment (mshriver) [reply robottelo/config/__init__.py:18]: I don't see there that `LazySettings` is or will be deprecated, just that its the default alias.

That's why we use it explicitly here. If Dynaconf upstream wants to change the default alias, our object would change without our knowledge.

File: scripts/vault_login.py
Comment (mshriver) [inline scripts/vault_login.py:90]: I feel like this is written in bash, in python :) 

Click is really good for creating extensible command structures. This file is largely a convenience - the only variable necessary to be able to use vault is the token, and to be able to use `vault` on the command line the URL.

I don't want us mirroring the vault CLI instead of using both `VAULT_ADDR` and `VAULT_URL_FOR_DYNACONF` in the `.env.example` file. 

That file is read by dynaconf, but it also can be sourced in one's local env in order to just use `vault` CLI directly and have the URL set.

Having a tool that makes integrating with the vault CLI is helpful, but I'm not sure this is the right approach. Using python's entrypoints and command groups implemented through click creates a far more extensible interface for doing things like this to 'work' with the framework and setup a testing environment.

File: scripts/vault_login.py
Comment (mshriver) [inline scripts/vault_login.py:80]: These checks against returncode are redundant, `_vault_command` will `sys.exit()` if its non-zero.

File: scripts/vault_login.py
Comment (mshriver) [inline scripts/vault_login.py:44]: These are effectively 'magic' numbers and are really annoying to work with. constants should be defined to describe what they are.

File: scripts/vault_login.py
Comment (mshriver) [inline scripts/vault_login.py:51]: 0 exit code is for a successful result, I'd suggest using something else

File: scripts/vault_login.py
Comment (mshriver) [inline scripts/vault_login.py:62]: Why use `os.system` here and `subprocess.run` above?

File: scripts/vault_login.py
Comment (mshriver) [inline scripts/vault_login.py:60]: You really should use `input` here instead of a static 5s sleep. it could easily take the user longer than that to complete OIDC if its not already up.

**Review by mshriver — CHANGES_REQUESTED** [review]
I think the vault login script needs some discussion, but the rest is looking good - lots of cleanup to the validators, thanks!

File: scripts/vault_login.py
Comment (jyejare) [reply scripts/vault_login.py:90]: Ofcorse there are multiple ways of implementing things, but since this is just a helper module to be used with `Make`, I kept it simple! 

I would not clock the PR on this helper module improvements/re-implementation, we can merge this PR and later one can raise a PR to improve it !

File: scripts/vault_login.py
Comment (jyejare) [reply scripts/vault_login.py:80]: Removed the check!

File: scripts/vault_login.py
Comment (jyejare) [reply scripts/vault_login.py:44]: Not a blocker for this PR and defining constants won't be reused outside of this module! Lets keep it simple!

File: scripts/vault_login.py
Comment (jyejare) [reply scripts/vault_login.py:51]: Replaced with 1, mistakenly kept it !

File: scripts/vault_login.py
Comment (jyejare) [reply scripts/vault_login.py:62]: Forgot to replace :)

Updated !

File: scripts/vault_login.py
Comment (jyejare) [reply scripts/vault_login.py:60]: The browser will open post this sleep. The sleep is for a user to not surprise why the browser is opened or just to inform that `vault login -method=oidc` will open the browser ! 

The vault login is not async , so the execution wont proceed to the next line until it finishes OIDC login.

File: scripts/vault_login.py
Comment (jyejare) [reply scripts/vault_login.py:44]: But still kept updated :)

**Comment by jyejare** [comment]
@mshriver Pushed changes to vault login script !

File: scripts/vault_login.py
Comment (mshriver) [reply scripts/vault_login.py]: I would consider this a blocker. These formatting codes are used many times in this module, and make it kinda hard to read what the messages are, and what formatting is being applied.

Comment (jyejare) [reply scripts/vault_login.py]: @mshriver I updated the code already!

**Review by mshriver — APPROVED** [review]
I've expressed some concerns with the structure of the bash scripts for vault login integration, but am not going to block merge on that.

Thanks @jyejare for being willing to pivot away from the initial proposed implementation to address my concerns with hardcoding paths into the validators!

**Comment by mshriver** [comment]
@jyejare time for a backport to 6.10.z!

**Pull request merged**
